File: P1368_最小表示法.py
from typing import List

def get_min_representation(s: List[int]) -> int:
    N = len(s)
    i = 0
    j = 1
    while i < N and j < N:
        k = 0
        while k < N and s[(i + k) % N] == s[(j + k) % N]:
            k += 1
        if k == N:
            break
        if s[(i + k) % N] < s[(j + k) % N]:
            j += k + 1
            if i == j:
                j += 1
        else:
            i += k + 1
            if i == j:
                i += 1
    return min(i, j)

def main():
    N = int(input())
    s = [int(x) for x in input().split()]
    ans = get_min_representation(s)

    for k in range(N):
        print("{} ".format(s[(ans + k) % N]), end="")


if __name__ == "__main__":
    main()

File: P1368_最小表示法.cpp
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    int N;
    cin >> N;
    vector<int> s(N);
    for(int i = 0; i < N; ++i)
        cin >> s[i];

    int i = 0, j = 1;
    while(i < N && j < N)
    {
        int k = 0;
        while(k < N && s[(i + k) % N] == s[(j + k) % N])
            k++;
        if(k == N)
            break;
        if(s[(i + k) % N] < s[(j + k) % N])
        {
            j += k + 1;
            if(i == j)
                j++;
        }
        else
        {
            i += k + 1;
            if(i == j)
                i++;
        }
    }
    int ans = min(i, j);
    for(int k = 0; k < N; ++k)
        cout << s[(ans + k) % N] << " ";
    cout << endl;
}
